File: source/python/avl_tree.py
class Node:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None
        self.height = 1

def height(node):
    return node.height if node else 0

def get_balance(node):
    return height(node.left) - height(node.right) if node else 0

def right_rotate(y):
    x = y.left
    t2 = x.right
    x.right = y
    y.left = t2
    y.height = max(height(y.left), height(y.right)) + 1
    x.height = max(height(x.left), height(x.right)) + 1
    return x

def left_rotate(x):
    y = x.right
    t2 = y.left
    y.left = x
    x.right = t2
    x.height = max(height(x.left), height(x.right)) + 1
    y.height = max(height(y.left), height(y.right)) + 1
    return y

def insert(root, value):
    if root is None:
        return Node(value)
    if value < root.value:
        root.left = insert(root.left, value)
    elif value > root.value:
        root.right = insert(root.right, value)
    else:
        return root
    root.height = max(height(root.left), height(root.right)) + 1
    balance = get_balance(root)
    if balance > 1 and value < root.left.value:
        return right_rotate(root)
    if balance < -1 and value > root.right.value:
        return left_rotate(root)
    if balance > 1 and value > root.left.value:
        root.left = left_rotate(root.left)
        return right_rotate(root)
    if balance < -1 and value < root.right.value:
        root.right = right_rotate(root.right)
        return left_rotate(root)
    return root

def inorder(root, result):
    if root is None:
        return
    inorder(root.left, result)
    result.append(root.value)
    inorder(root.right, result)

root = None
for v in [10, 20, 30, 40, 50, 25]:
    root = insert(root, v)

result = []
inorder(root, result)
print(f"inorder: {' '.join(str(x) for x in result)}")

File: source/javascript/avl_tree.js
// AVL tree with rotations
function newNode(value) { return { value, left: null, right: null, height: 1 }; }

function height(node) { return node === null ? 0 : node.height; }
function maxVal(a, b) { return a > b ? a : b; }
function getBalance(node) { return node === null ? 0 : height(node.left) - height(node.right); }

function rightRotate(y) {
    const x = y.left, t2 = x.right;
    x.right = y;
    y.left = t2;
    y.height = maxVal(height(y.left), height(y.right)) + 1;
    x.height = maxVal(height(x.left), height(x.right)) + 1;
    return x;
}

function leftRotate(x) {
    const y = x.right, t2 = y.left;
    y.left = x;
    x.right = t2;
    x.height = maxVal(height(x.left), height(x.right)) + 1;
    y.height = maxVal(height(y.left), height(y.right)) + 1;
    return y;
}

function insert(root, value) {
    if (root === null) return newNode(value);
    if (value < root.value) root.left = insert(root.left, value);
    else if (value > root.value) root.right = insert(root.right, value);
    else return root;

    root.height = maxVal(height(root.left), height(root.right)) + 1;
    const balance = getBalance(root);

    if (balance > 1 && value < root.left.value) return rightRotate(root);
    if (balance < -1 && value > root.right.value) return leftRotate(root);
    if (balance > 1 && value > root.left.value) { root.left = leftRotate(root.left); return rightRotate(root); }
    if (balance < -1 && value < root.right.value) { root.right = rightRotate(root.right); return leftRotate(root); }
    return root;
}

function inorder(root, result) {
    if (root === null) return;
    inorder(root.left, result);
    result.push(root.value);
    inorder(root.right, result);
}

let root = null;
const values = [10, 20, 30, 40, 50, 25];
for (const v of values) root = insert(root, v);

const result = [];
inorder(root, result);
console.log(`inorder: ${result.join(" ")}`);
